### PR TITLE
feat: adds conversion of softmax from onnx to tosa as well as utilities

### PR DESCRIPTION
**Pull request by maxbartel**
Softmax does not have a 1:1 conversion to TOSA. To lower it we have to decompose it. The shape inference builder that TensorFlow and Pytorch are using was also added.

Signed-off-by: Maximilian Bartel <maximilian.bartel@amd.com>

### CODE REVIEW

File: src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//==== ONNXToTosaLegalizeUtils.hpp - ONNX dialects to TOSA lowering Utils-===//
+//
+// Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp:8]: I'm not suggesting we already change if there aren't any questions, but Luis recently pointed me to an example (https://github.com/amd/UIF/blob/main/samples/tf_sample.py) which we could implement here as something like:
```
// Copyright 2020 The TensorFlow Authors. All Rights Reserved.
//
// Modifications notified per clause 4(b) of the license:
// Copyright (c) 2022 Advanced Micro Devices, Inc.
```

File: src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+// This file contains common utils shared by the functions performing the
+// lowering to the TOSA dialect. It is also used by TensorFlow and Pytorch.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef ONNXMLIR_CONVERSION_ONNXTOTOSA_TOSALEGALIZEUTILS_H
+#define ONNXMLIR_CONVERSION_ONNXTOTOSA_TOSALEGALIZEUTILS_H
+
+#include "mlir/Dialect/Quant/QuantTypes.h" // from @llvm-project
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp:21]: Is the `// from @llvm-project` annotation useful? And if so, why isn't it on `"mlir/Dialect/Tosa/IR/TosaOps.h"`?

Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp:21]: Good question, I think this is a relic from adopting code from TF. I will delete the comments

File: src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp (hunk 1 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp:8]: So we leave it like this until someone complains?

**Review by ljfitz — APPROVED** [review]

File: src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+//==== ONNXToTosaLegalizeUtils.hpp - ONNX dialects to TOSA lowering Utils-===//
+//
+// Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file contains common utils shared by the functions performing the
+// lowering to the TOSA dialect. It is also used by TensorFlow and Pytorch.
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp:13]:
```suggestion
// lowering to the TOSA dialect. It is also used by TensorFlow and torch-mlir.
```
